sql/core/src/main/scala/org/apache/spark/sql/DataFrameWriter.scala

```diff
@@ -218,7 +247,14 @@ final class DataFrameWriter[T] private[sql](ds: Dataset[T]) {
       bucketSpec = getBucketSpec,
       options = extraOptions.toMap)
 
-    dataSource.write(mode, df)
+    val destination = source match {
+      case "jdbc" => extraOptions.get(JDBCOptions.JDBC_TABLE_NAME)
+      case _ => extraOptions.get("path")
```

Comment:
So that comment is about different types of queries that might have different extra information. I still think a generic map is the wrong idea for fixing that.

You could, for example, have this parameter be `Any` (or some tagging interface, e.g. `trait QueryExecutionParams` or some such), and the listener can then match to find the right type. That is extensible (new types can be added without breaking existing ones) and wouldn't require this API to change in the future.